<hr>
<font color='purple'>
    <font size="+3"><strong> 
Name: Theophilus Uwoghiren
</font>

<br>
<hr>

<font color='green'>
    <font size="+3"><strong> 
Hamoye ID: 14843ce775c1f000
 </font>

<br>
<hr>

<font color='black'>
    <font size="+3">
Track: Data Science 
</font>         

<br>
<hr>
    
<font size="+3">
Stage A Assessment
</font>
<hr>

<br>
<font size="+3"><strong> 
Data Preprocessing
</strong>
</font>

**Importing Libraries**

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

**Reading the Dataset**

In [2]:
df = pd.read_csv('FoodBalanceSheets_E_Africa_NOFLAG.csv', encoding="latin-1")

df.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

**QUESTION 1**

What is the total sum of Wine produced in 2015 and 2018 respectively?

In [3]:
df.isin(['Wine']).any()

Area Code       False
Area            False
Item Code       False
Item             True
Element Code    False
Element         False
Unit            False
Y2014           False
Y2015           False
Y2016           False
Y2017           False
Y2018           False
dtype: bool

In [4]:
df.groupby("Item")[["Y2015","Y2018"]].sum()

Y2015      Y2018
Item                                     
Alcohol, Non-Food      2180.00    2293.00
Alcoholic Beverages   98783.72   97847.27
Animal Products       11811.73   11578.61
Animal fats          200675.72  269648.27
Apples and products   10559.15    9640.51
...                        ...        ...
Vegetables, Other    158104.08  163987.21
Vegetal Products     107064.17  107775.39
Wheat and products   234710.51  242645.19
Wine                   4251.81    4039.32
Yams                 203151.78  221272.09

[119 rows x 2 columns]

**Answer:** The total sum of Wine produced in 2015 and 2018 respectively is **4251.81** and **4039.32**

**QUESTION 2**

Given the following python code, what would the output of the code give?

my_tuppy = (1,2,5,8)

my_tuppy[2] = 6


**Answer:** Type Error

**QUESTION 3**

Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?

In [5]:
df.groupby("Area")["Y2018"].sum().sort_values(ascending=False)

Area
Nigeria                        1499774.24
Egypt                           852652.92
South Africa                    509787.97
Ethiopia                        450788.86
Morocco                         399271.27
Ghana                           358915.64
United Republic of Tanzania     349763.41
Algeria                         341729.98
Kenya                           270916.80
Sudan                           268054.39
Angola                          235206.68
Cameroon                        233022.93
C�te d'Ivoire                 227018.00
Uganda                          222147.25
Malawi                          187366.09
Mozambique                      164934.05
Mali                            159026.42
Mauritania                      157479.95
Madagascar                      134576.44
Niger                           133982.86
Tunisia                         127343.17
Benin                           124960.70
Burkina Faso                    112970.72
Guinea                       

**Answer:** Nigeria

In [6]:
df.shape

(60943, 12)

**QUESTION 4**


What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

In [7]:
df['Y2017'].describe().round(2)

count     59437.00
mean        140.92
std        1671.86
min       -1582.00
25%           0.00
50%           0.10
75%           9.00
max      190873.00
Name: Y2017, dtype: float64

**Answer:** The mean is **140.92** while the standard deviation is **1671.86**

**QUESTION 8**

What is the total number of unique countries in the dataset?

In [8]:
df['Area'].nunique()

49

**Answer**:  The total number of unique countries in the dataset is 49

**QUESTION 9**


Which year had the least correlation with ‘Element Code’? 

In [9]:
F_Correlation = df.corr()
F_Correlation['Element Code'].sort_values(ascending=True)

Item Code      -0.024683
Area Code      -0.000209
Y2016           0.023444
Y2015           0.023889
Y2017           0.024254
Y2018           0.024279
Y2014           0.024457
Element Code    1.000000
Name: Element Code, dtype: float64

**Answer**: Year **2016** has the least correlation with ‘Element Code'

**QUESTION 10**

What would be the output for?

In [10]:
S = [['him', 'sell'], [90, 28, 43]]

S[0][1][1]

'e'

**Answer**: 'e'

**QUESTION 11**

Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017?

In [11]:
mask_df = df['Element'].str.contains('Processing')
df[mask_df] 


Area Code      Area  Item Code                      Item  Element Code  \
19             4   Algeria       2905  Cereals - Excluding Beer          5131   
35             4   Algeria       2511        Wheat and products          5131   
51             4   Algeria       2805         Rice and products          5131   
66             4   Algeria       2513       Barley and products          5131   
81             4   Algeria       2514        Maize and products          5131   
...          ...       ...        ...                       ...           ...   
60694        181  Zimbabwe       2946               Animal fats          5131   
60717        181  Zimbabwe       2743                     Cream          5131   
60730        181  Zimbabwe       2737        Fats, Animals, Raw          5131   
60789        181  Zimbabwe       2948   Milk - Excluding Butter          5131   
60803        181  Zimbabwe       2848   Milk - Excluding Butter          5131   

          Element         Unit  Y2014  Y2015  Y2016  Y2017  Y2018  
19     Processing  1000 tonnes   35.0   38.0   37.0   33.0   44.0  
35     Processing  1000 tonnes    0.0    0.0    0.0    0.0    0.0  
51     Processing  1000 tonnes    0.0    0.0    0.0    0.0    0.0  
66     Processing  1000 tonnes   35.0   38.0   37.0   33.0   44.0  
81     Processing  1000 tonnes    0.0    0.0    0.0    0.0    0.0  
...           ...          ...    ...    ...    ...    ...    ...  
60694  Processing  1000 tonnes    0.0    0.0    0.0    0.0    0.0  
60717  Processing  1000 tonnes    0.0    0.0    0.0    0.0    0.0  
60730  Processing  1000 tonnes    0.0    0.0    0.0    0.0    0.0  
60789  Processing  1000 tonnes   59.0   59.0   58.0   56.0   57.0  
60803  Processing  1000 tonnes   59.0   59.0   58.0   56.0   57.0  

[2010 rows x 12 columns]

In [12]:
df.isin(['Processing']).any()

Area Code       False
Area            False
Item Code       False
Item            False
Element Code    False
Element          True
Unit            False
Y2014           False
Y2015           False
Y2016           False
Y2017           False
Y2018           False
dtype: bool

In [13]:
Q11=df.groupby('Element')['Y2017'].sum()
Q11

Element
Domestic supply quantity                  2088198.10
Export Quantity                            182338.80
Fat supply quantity (g/capita/day)          10253.84
Feed                                       223705.68
Food                                      1258888.28
Food supply (kcal/capita/day)              454681.00
Food supply quantity (kg/capita/yr)         48690.04
Import Quantity                            294559.09
Losses                                     160614.00
Other uses (non-food)                       91645.97
Processing                                 292836.00
Production                                2030056.89
Protein supply quantity (g/capita/day)      11842.45
Residuals                                   35500.00
Seed                                        24870.14
Stock Variation                             54316.91
Total Population - Both sexes             1112641.00
Tourist consumption                            91.00
Name: Y2017, dtype: float64

**Answer**: The total number of the sum of Processing in 2017 is **292836.00**

**Question 13**

Perform a groupby operation on ‘Element’.  What year has the highest sum of Stock Variation?


In [14]:
df.isin(['Import Quantity']).any()

Area Code       False
Area            False
Item Code       False
Item            False
Element Code    False
Element          True
Unit            False
Y2014           False
Y2015           False
Y2016           False
Y2017           False
Y2018           False
dtype: bool

In [15]:
Q13 = df.groupby('Element')[['Y2014','Y2015','Y2016','Y2017','Y2018']].sum()
Q13


Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00   280631.00   
Production                              1931287.75  1947019.39  1943537.15   
Protein supply quantity (g/capita/day)    11836.46    11833.95    11779.69   
Residuals                                 30149.00    30045.00    37224.00   
Seed                                      21922.92    23976.82    23389.20   
Stock Variation                           58749.83    34910.99    33140.12   
Total Population - Both sexes           1031585.00  1058081.00  1085107.00   
Tourist consumption                         416.00      349.00       89.00   

                                             Y2017       Y2018  
Element                                                         
Domestic supply quantity                2088198.10  2161192.10  
Export Quantity                          182338.80   181594.80  
Fat supply quantity (g/capita/day)        10253.84    10258.69  
Feed                                     223705.68   233489.68  
Food                                    1258888.28  1303841.28  
Food supply (kcal/capita/day)            454681.00   455261.00  
Food supply quantity (kg/capita/yr)       48690.04    49056.85  
Import Quantity                          294559.09   287997.09  
Losses                                   160614.00   163902.00  
Other uses (non-food)                     91645.97    91300.97  
Processing                               292836.00   308429.00  
Production                              2030056.89  2075072.89  
Protein supply quantity (g/capita/day)    11842.45    11833.56  
Residuals                                 35500.00    34864.00  
Seed                                      24870.14    25263.14  
Stock Variation                           54316.91    20577.91  
Total Population - Both sexes           1112641.00  1140605.00  
Tourist consumption                          91.00       90.00

In [16]:
Q13.loc['Stock Variation']


Y2014    58749.83
Y2015    34910.99
Y2016    33140.12
Y2017    54316.91
Y2018    20577.91
Name: Stock Variation, dtype: float64

In [17]:
Q13.loc['Stock Variation'].at['Y2014'].round(2)

58749.83

**Answer:** The highest sum of Stock Variation is **58749.83**

**Question 15**

Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the 7th lowest sum in 2017?


In [18]:
df.groupby("Area")["Y2017"].sum().sort_values(ascending=True)

Area
Sudan (former)                       0.00
Ethiopia PDR                         0.00
Comoros                             59.84
Seychelles                         442.34
Sao Tome and Principe            12662.63
Cabo Verde                       14650.74
Guinea-Bissau                    19102.77
Lesotho                          21267.96
Botswana                         22101.30
Djibouti                         22729.91
Gambia                           23154.18
Gabon                            27979.64
Liberia                          29342.20
Namibia                          29874.89
Central African Republic         29937.00
Congo                            41181.68
Togo                             49841.88
Mauritius                        51114.83
Eswatini                         54343.33
Sierra Leone                     55311.33
Chad                             71594.68
Rwanda                           73663.69
Zimbabwe                         75919.34
Senegal                      

**Answer**: The area with the 7th lowest sum in 2017 is **Guinea-Bissau**

**Question 18**

What is the total number and percentage of missing data in 2014 to 3 decimal places?


In [19]:
df['Y2014'].isnull().sum()

1589

In [20]:
df['Y2014'].count()

59354

In [21]:
Percent_Missing_Data = (df['Y2014'].isnull().sum())*100/(len(df['Y2014']))
Answer = round(Percent_Missing_Data,3)

In [22]:
print("{}%".format(Answer))

2.607%


**Answer:** 
    The total number of missing data in 2014 is **1589**;
    The percentage of missing data in 2014 to 3 decimal places is **2.607%**

**Question 19**

What is the total Protein supply quantity in Madagascar in 2015?


In [23]:
Q19 = df.pivot_table(index=['Area','Element'],values =['Y2014', 'Y2015', 'Y2016', 'Y2017', 'Y2018'],aggfunc='sum')
Q19

Y2014     Y2015     Y2016  \
Area     Element                                                                
Algeria  Domestic supply quantity                83755.94  85972.86  87157.61   
         Export Quantity                          1328.64   1104.62   1340.22   
         Fat supply quantity (g/capita/day)        320.19    317.40    309.54   
         Feed                                     9656.04  10308.00   9903.00   
         Food                                    60582.69  61693.49  63081.43   
...                                                   ...       ...       ...   
Zimbabwe Protein supply quantity (g/capita/day)    184.23    175.31    172.72   
         Residuals                                -252.00   -780.00   -882.00   
         Seed                                      144.00    128.00    127.00   
         Stock Variation                          -644.00   -343.00    221.00   
         Total Population - Both sexes           13587.00  13815.00  14030.00   

                                                    Y2017     Y2018  
Area     Element                                                     
Algeria  Domestic supply quantity                86810.42  90023.42  
         Export Quantity                          1494.54    811.54  
         Fat supply quantity (g/capita/day)        320.74    309.09  
         Feed                                     9571.00  10668.00  
         Food                                    62785.33  63736.33  
...                                                   ...       ...  
Zimbabwe Protein supply quantity (g/capita/day)    175.31    176.57  
         Residuals                               -1083.00   -546.00  
         Seed                                      162.00    140.00  
         Stock Variation                           458.00   -447.00  
         Total Population - Both sexes           14237.00  14439.00  

[803 rows x 5 columns]

In [24]:
Q19.loc['Madagascar','Y2015']

Element
Domestic supply quantity                  31214.98
Export Quantity                             494.75
Fat supply quantity (g/capita/day)           91.85
Feed                                       2070.22
Food                                      21120.65
Food supply (kcal/capita/day)              7685.00
Food supply quantity (kg/capita/yr)         871.59
Import Quantity                            1721.80
Losses                                     2176.00
Other uses (non-food)                      3957.12
Processing                                 1803.00
Production                                29482.89
Protein supply quantity (g/capita/day)      173.05
Residuals                                  -616.00
Seed                                        699.00
Stock Variation                            -505.00
Total Population - Both sexes             24234.00
Name: Y2015, dtype: float64

**Answer:** Protein supply quantity in Madagascar in 2015 is **173.05**g/capita/day